crates/lang/codegen/src/generator/selector.rs

```diff
@@ -0,0 +1,55 @@
+// Copyright 2018-2021 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::GenerateCode;
+use derive_more::From;
+use ir::HexLiteral;
+use proc_macro2::TokenStream as TokenStream2;
+use quote::quote_spanned;
+
+/// Generates code for the `selector_id!` macro.
+#[derive(From)]
+pub struct SelectorId<'a> {
+    /// The contract to generate code for.
+    macro_input: &'a ir::SelectorMacro<ir::marker::SelectorId>,
```

Comment:
Can you explain why we require the marker type parameter for `SelectorMacro`?

Reply:
this is just to avoid some code dupe because both `SelectorId` and `SelectorBytes` are using the same underlying mechanics.